### PR TITLE
Redefine Foundry test for LONStaking test contracts

### PR DESCRIPTION
**Pull request by oneleo**
### Rules:

1. One file, one contract.
2. Test contracts are named with upper camel case.
3. Test functions are named with lower camel case, and reduce name according to functions' content.
4. Minimal Import principle: do not import unused libraries.
5. Minimal Using/Struct/Helpers principle.
6. Use MIT SPDX-License-Identifier.

Please use following command to test:
```
% forge test -vvv --match-path "contracts/test/LONStaking/*.t.sol" --force
```

The execution result is as follows:
<details>
  <summary>( Please click here to expand )</summary>
    <pre><code>

Running 4 tests for contracts/test/LONStaking/Unstake.t.sol:TestLONStakingUnstake
[PASS] testCannotUnstakeAgain() (gas: 167672)
[PASS] testCannotUnstakeWithZeroAmount() (gas: 20537)
[PASS] testUnstake() (gas: 168885)
[PASS] testUnstakeWhenPaused() (gas: 176954)
Test result: ok. 4 passed; 0 failed; finished in 23.72ms

Running 4 tests for contracts/test/LONStaking/Pause.t.sol:TestLONStakingPause
[PASS] testCannotPauseByNotOwner() (gas: 20438)
[PASS] testCannotUnpauseByNotOwner() (gas: 28491)
[PASS] testPause() (gas: 23135)
[PASS] testUnpause() (gas: 23652)
Test result: ok. 4 passed; 0 failed; finished in 26.35ms

Running 3 tests for contracts/test/LONStaking/SetCooldownAndRageExitParam.t.sol:TestLONStakingSetCooldownAndRageExitParam
[PASS] testCannotSetByNotOwner() (gas: 20530)
[PASS] testCannotSetWithInvalidParam() (gas: 28862)
[PASS] testSetCooldownAndRageExitParam() (gas: 34988)
Test result: ok. 3 passed; 0 failed; finished in 30.82ms

Running 3 tests for contracts/test/LONStaking/RecoverERC20.t.sol:TestLONStakingRecoverERC20
[PASS] testCannotRecoverByNotOwner() (gas: 22665)
[PASS] testCannotRecoverWithLONToken() (gas: 22374)
[PASS] testRecoverERC20() (gas: 849770)
Test result: ok. 3 passed; 0 failed; finished in 14.28ms

Running 4 tests for contracts/test/LONStaking/SetupTest.t.sol:TestLONStakingSetup
[PASS] testCannotReinitialize() (gas: 24674)
[PASS] testCannotUpgradeProxyByNotAdmin() (gas: 20559)
[PASS] testLONStakingSetup() (gas: 28876)
[PASS] testUpgradeProxy() (gas: 25511)
Test result: ok. 4 passed; 0 failed; finished in 63.40ms

Running 5 tests for contracts/test/LONStaking/StakeWithPermit.t.sol:TestLONStakingStakeWithPermit
[PASS] testCannotStakeWhenPaused() (gas: 71663)
[PASS] testCannotStakeWhenPermitExpired() (gas: 116937)
[PASS] testCannotStakeWhenSignatureSeenBefore() (gas: 193448)
[PASS] testCannotStakeWithInvalidSignature() (gas: 144983)
[PASS] testStakeWithPermit() (gas: 193879)
Test result: ok. 5 passed; 0 failed; finished in 12.30ms

Running 11 tests for contracts/test/LONStaking/RageExit.t.sol:TestLONStakingRageExit
[PASS] testCannotExitBeforeUnstake() (gas: 42586)
[PASS] testFuzz_RageExitOneByOnePlusPenaltyWithMultipleStake(uint80[16]) (runs: 1000, μ: 2599248, ~: 2595120)
[PASS] testFuzz_RageExitOneByOneWithBuybackPlusPenaltyWithMultipleStake(uint80[16],uint80[16]) (runs: 1000, μ: 2405050, ~: 2411968)
[PASS] testFuzz_RageExitOneTimePlusPenaltyWithMultipleStake(uint80[16]) (runs: 1000, μ: 2308501, ~: 2308502)
[PASS] testFuzz_RageExitOneTimeWithBuybackPlusPenaltyWithMultipleStake(uint80[16],uint80) (runs: 1000, μ: 2313322, ~: 2313323)
[PASS] testFuzz_RageExitWithPenalty(uint256) (runs: 1000, μ: 173909, ~: 173914)
[PASS] testPreviewRageExitWithPenalty() (gas: 174675)
[PASS] testPreviewRageExitWithoutPenalty() (gas: 169233)
[PASS] testRageExitWithBuybackPlusPenalty() (gas: 169958)
[PASS] testRageExitWithPenalty() (gas: 161279)
[PASS] testRageExitWithoutPenalty() (gas: 156004)
Test result: ok. 11 passed; 0 failed; finished in 182.69s

Running 7 tests for contracts/test/LONStaking/Stake.t.sol:TestLONStakingStake
[PASS] testCannotStakeWhenPaused() (gas: 52821)
[PASS] testCannotStakeWithZeroAmount() (gas: 40115)
[PASS] testFuzz_Stake(uint256) (runs: 1000, μ: 163972, ~: 163972)
[PASS] testFuzz_StakeMultiple(uint80[16]) (runs: 1000, μ: 1505973, ~: 1505973)
[PASS] testFuzz_StakeMultipleWithBuybackOneByOne(uint80[16],uint80[16]) (runs: 1000, μ: 1657675, ~: 1660481)
[PASS] testFuzz_StakeMultipleWithBuybackOneTime(uint80[16],uint80) (runs: 1000, μ: 2999371, ~: 3003152)
[PASS] testStake() (gas: 153787)
Test result: ok. 7 passed; 0 failed; finished in 192.51s

Running 5 tests for contracts/test/LONStaking/Permit.t.sol:TestLONStakingPermit
[PASS] testCannotPermitByZeroAddress() (gas: 33967)
[PASS] testCannotPermitWhenPermitExpired() (gas: 34036)
[PASS] testCannotPermitWhenSignatureSeenBefore() (gas: 92685)
[PASS] testCannotPermitWithInvalidSignature() (gas: 62266)
[PASS] testPermit() (gas: 96252)
Test result: ok. 5 passed; 0 failed; finished in 216.29s

Running 17 tests for contracts/test/LONStaking/Redeem.t.sol:TestLONStakingRedeem
[PASS] testCannotRedeemAgain() (gas: 144084)
[PASS] testCannotRedeemBeforeCooldown() (gas: 171895)
[PASS] testCannotRedeemBeforeUnstake() (gas: 42725)
[PASS] testCannotRedeemMoreThanOwned() (gas: 175086)
[PASS] testCannotRedeemWithZeroAmount() (gas: 170266)
[PASS] testFuzz_RedeemAll(uint256) (runs: 1000, μ: 171305, ~: 171310)
[PASS] testFuzz_RedeemAllOneByOneWithMultipleStake(uint80[16]) (runs: 1000, μ: 2791277, ~: 2791279)
[PASS] testFuzz_RedeemAllOneTimeWithMultipleStake(uint80[16]) (runs: 1000, μ: 2313475, ~: 2313476)
[PASS] testFuzz_RedeemMultipleOneByOneWithBuyback(uint80[2][16],uint80[16]) (runs: 1000, μ: 2770894, ~: 2778274)
[PASS] testFuzz_RedeemMultipleOneTimeWithBuyback(uint80[2][16],uint80) (runs: 1000, μ: 2614441, ~: 2614461)
[PASS] testFuzz_RedeemPartial(uint256,uint256) (runs: 1000, μ: 217309, ~: 217309)
[PASS] testFuzz_RedeemPartialOneByOneWithMultipleStake(uint80[2][16]) (runs: 1000, μ: 2611600, ~: 2611610)
[PASS] testFuzz_RedeemPartialOneTimeWithMultipleStake(uint80[2][16]) (runs: 1000, μ: 2608402, ~: 2608412)
[PASS] testRedeemAll() (gas: 163260)
[PASS] testRedeemPartial() (gas: 205548)
[PASS] testRedeemWhenPaused() (gas: 191000)
[PASS] testRedeemWithBuyback() (gas: 170258)
Test result: ok. 17 passed; 0 failed; finished in 216.29s
    </code></pre>
</details>

Thank you very much!

### CODE REVIEW

**Review by NIC619 — DISMISSED** [review]

File: contracts/test/LONStaking/Setup.t.sol [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    function _stake(address staker, uint256 stakeAmount) internal {
+        vm.startPrank(staker);
+        if (lon.allowance(staker, address(lonStaking)) == 0) {
+            lon.approve(address(lonStaking), type(uint256).max);
+        }
+        lonStaking.stake(stakeAmount);
+        vm.stopPrank();
+    }
+
+    function simulateBuyback(uint256 amount) internal {
```
Comment (pilagod) [inline contracts/test/LONStaking/Setup.t.sol:97]: Shall we prefix underscore to this internal function too?

Comment (oneleo) [reply contracts/test/LONStaking/Setup.t.sol:97]: Yes! Its original without underscore, but it should need to be added.

Comment (charlesjhongc) [reply contracts/test/LONStaking/Setup.t.sol:97]: So is this PR ready or need adjusted?

Comment (oneleo) [reply contracts/test/LONStaking/Setup.t.sol]: Sorry, I have already pushed a new commit!
Waiting for the CI program...

**Review by charlesjhongc — APPROVED** [review]

**Review by NIC619 — APPROVED** [review]

**Review by pilagod — APPROVED** [review]

**Pull request merged**